### ISSUE TITLE
[Feature Request] Documentation explaining the recommended way to save and load agents

### ISSUE THREAD

**Issue by valterschutz** [pytorch/rl#3044]
## Motivation

I would like a concrete documented example of how to save RL agents after they have been trained, and also how to load them afterwards to resume training. In my current projects, I always switch between different methods and I have not yet found an approach that feels "right". For example,
- Should only the `nn.Module`s be saved or the `TensorDictModule`s?
- Should the loss module be saved?
- Saving the state_dict is usually recommended for `nn.Module`s. Is it the same way here?

I know that https://docs.pytorch.org/rl/stable/tutorials/export.html already explains how to export only the policy for inference, but I'm looking for something more general which would allow to resume training of the agent.

 #2620 and #1089 are related.

## Solution

I'm proposing that one of the following pages gets added to the documentation:
1. A separate page that describes how to save a DQN agent (or some other off-policy method). Network weights, optimizer state  *and* the replay buffer should be saved so that training can resume.
2. In the PPO tutorial (https://docs.pytorch.org/rl/stable/tutorials/coding_ppo.html#), add a section at the end that explains how to save the trained agent. The drawback with this approach is that it doesn't show how to save a replay buffer.

## Alternatives

I don't think there is. Saving models after they have been trained is an essential part of the RL workflow and the recommended way to do this should definitely be documented somewhere.

## Checklist

- [X] I have checked that there is no similar issue in the repo (**required**)

**Comment by vmoens** [comment]
You're right — maybe this is a good time to think out loud about what an ideal checkpointing API should look like.

At a high level, resuming training in RL usually requires saving the following components:

* **Training setup**

  * The training model
  * Optimizer
  * Target networks

* **Collector**

  * The inference model (possibly distinct from the training one), including any exploration coefficients
  * The environment state, including transforms
  * The collector state (e.g., how many frames have been collected)

* **Replay buffer**

  * Writer
  * Sampler
  * Transforms
  * Underlying storage

Each of these has its own quirks, and like you pointed out, things get more subtle in RL — for example, the training and inference models *might* share weights, but not always. That extra layer of complexity is something our checkpointing design should account for.

It also raises the question: how do we define what should be checkpointed *vs.* what should be exported?
In my mind, **checkpointing** is about interrupting training with the intent to resume it later or to inspect intermediate states. **Exporting**, on the other hand, is about saving a model for deployment — and in that context, trimming training-specific components *does* make sense. These goals don’t necessarily align, and we should make that distinction explicit. Still, we should be able to *convert* a checkpoint into an exportable artifact when needed.

As for the actual API, if we have a `Trainer` class that manages all components, checkpointing becomes much easier — we can handle the serialization internally and reduce the user's burden. Without such a wrapper, we'd need really solid documentation to help users correctly save and restore all the pieces.

Here’s one idea: what if we documented how to serialize each component individually — and then introduced a `Checkpointer` class to help users pull it all together? Something like:

```python
class Checkpointer:
    collector: DataCollectorBase | None = None
    training_agent: Any | None = None
    replay_buffer: ReplayBuffer | None = None
    training_env: EnvBase | None = None
    evaluation_env: EnvBase | None = None

    def checkpoint(self, path: str, **kwargs): ...
    
    @classmethod
    def load_checkpoint(cls, path: str, **kwargs) -> "Checkpointer": ...

    def export(self, path: str, **kwargs): ...
```

This could provide a structured approach while still allowing for flexibility. Of course, we’d still want solid docs — but I think it’s just as important that the API makes doing the right thing easy and hard to mess up.

Curious what you think about this breakdown — especially the separation between checkpointing and exporting. Would love to hear if you’ve thought about other components we should account for, or if you see potential pitfalls with this direction.

**Comment by valterschutz** [comment]
Wow, thank you so much for the comprehensive answer!

First of all, I definitely agree with the distinction of "saving" vs "exporting". For exporting, I've tried using `torch.export` as explained in (https://docs.pytorch.org/rl/stable/tutorials/export.html) but my models were unfortunately not supported by it. I've used `torch.save` directly on the policy `TensorDictModule`, which seems to work, but I don't know if this is recommended. Best practice in pytorch seems to be to avoid using `torch.save` directly on models and instead prefer `torch.save(model.state_dict())`. I wonder how this translates to torchrl.

My confusion here is very much related to the question raised in #3032 . Since I don't really understand where the weights are stored, I also don't know what to save!

Regarding the checkpointing, it looks very comprehensive and would probably be useful. But currently, I'm much more interested in "document how to serialize each component individually", especially how to serialize the agent itself.
